### ISSUE TITLE
Test log output is not separated using TF_LOG_PATH_MASK

### ISSUE THREAD

**Issue by slevenick** [hashicorp/terraform-plugin-sdk#922]
### SDK version

```
...
```

Tested with `github.com/hashicorp/terraform-plugin-sdk/v2 v2.10.1` and `2.12.0`

### Expected Behavior
Output logs should be in separate files based on the individual test being run.

### Actual Behavior
Individual files are created per test, but they do not contain only the logs for that test. Actual logs seem to be concentrated in a single file, or split seemingly randomly between different files. Some files only contain SDK-level logs for their particular test. For example, this test's HTTP interactions were logged to a different file, leaving only SDK-level logs: [debug-TestAccIAM2DenyPolicy_iamDenyPolicyUpdate.log](https://github.com/hashicorp/terraform-plugin-sdk/files/8393849/debug-TestAccIAM2DenyPolicy_iamDenyPolicyUpdate.log)

### Steps to Reproduce
Set TF_LOG_PATH_MASK to something similar to: `$PWD/debug-%s.log`.

Run multiple tests in parallel using the standard go test command

Examine the log files

**Comment by detro** [comment]
To start, a question: do you use https://pkg.go.dev/github.com/hashicorp/terraform-plugin-log/tflog or just plain `log` package?

**Comment by detro** [comment]
While I wait for the reply, I did try to reproduce this: producing `log`s inside tests and actual provider code, while using `TF_LOG_PATH_MASK`, worked without issues. The files would be nicely separated per test and the output I was expecting to see was there.

About the SDK level logs though, I confirm that by setting a `TF_LOG=DEBUG` level, I do get lots SDK-level logs. But I think this is to be expected, as the SDK does use `tfsdklog`, that is part of `terraform-plugin-log`.

I'll @bflad to chime in, as I don't know if this a "by-design" behaviour or not.

**Comment by slevenick** [comment]
We are using the plain log package. Are you running multiple tests in parallel and seeing the output separated correctly?

**Comment by detro** [comment]
So, as I didn't have at hand a repo I know well enough to test, I locally edited one for this experiment.

* repo: https://github.com/hashicorp/terraform-provider-tls
* I replaced all calls to `resource.Test` or `resource.UnitTest` with `resource.ParallelTest`: this of course makes many tests fail (as they aren't designed for parallalization) but we don't care for this purpose
* I have setup the `Makefile` to run the tests in the following way:

```Makefile
testacc:
	TF_LOG=DEBUG \
	TF_LOG_PATH_MASK=$HOME/Workspaces/hashicorp/terraform-provider-tls/logs/debug-%s.log \
	TF_ACC=1 \
		go test -v -cover -timeout 120m -parallel=10 ./...
```

With this setup, of course, a lot of tests started failing (expected) but all of them contain log statements that are circumstantial to the test of the specific file. It's hard to be 100% certain, but from a spot check all the files named `debug-TestAccX` (or `debug-TestX`) contained lines like:

```
2022-04-07T12:29:51.809+0100 [DEBUG] sdk.helper_resource: Creating sdkv2 provider instance: test_name=TestAccX
```
plus of course any logging specific to that test.

In short, I can't seem to reproduce this issue.

**Comment by slevenick** [comment]
Attaching my folder of test logs when running a subset of tests. Note the duplicate extensions, I had to zip it again to get around GitHub's file extension limits
[masks.zip.gpg.zip](https://github.com/hashicorp/terraform-plugin-sdk/files/8478325/masks.zip.gpg.zip)

These were collected when running tests with `TF_ACC_LOG_PATH` unset, `TF_LOG=DEBUG` and `TF_LOG_PATH_MASK=$PWD/masks/debug-%s.log`

Notice that some tests have no contents, or a single API request without the normally associated logs.

Stdout had:
```
=== RUN   TestAccDialogflowCXAgent_dialogflowcxAgentFullExample
=== PAUSE TestAccDialogflowCXAgent_dialogflowcxAgentFullExample
=== RUN   TestAccDialogflowCXEntityType_dialogflowcxEntityTypeFullExample
=== PAUSE TestAccDialogflowCXEntityType_dialogflowcxEntityTypeFullExample
=== RUN   TestAccDialogflowCXEnvironment_dialogflowcxEnvironmentFullExample
=== PAUSE TestAccDialogflowCXEnvironment_dialogflowcxEnvironmentFullExample
=== RUN   TestAccDialogflowCXFlow_dialogflowcxFlowFullExample
=== PAUSE TestAccDialogflowCXFlow_dialogflowcxFlowFullExample
=== RUN   TestAccDialogflowCXIntent_dialogflowcxIntentFullExample
=== PAUSE TestAccDialogflowCXIntent_dialogflowcxIntentFullExample
=== RUN   TestAccDialogflowCXPage_dialogflowcxPageFullExample
=== PAUSE TestAccDialogflowCXPage_dialogflowcxPageFullExample
=== RUN   TestAccDialogflowCXVersion_dialogflowcxVersionFullExample
=== PAUSE TestAccDialogflowCXVersion_dialogflowcxVersionFullExample
=== RUN   TestAccDialogflowCXAgent_update
=== PAUSE TestAccDialogflowCXAgent_update
=== RUN   TestAccDialogflowCXEntityType_update
=== PAUSE TestAccDialogflowCXEntityType_update
=== RUN   TestAccDialogflowCXEnvironment_update
=== PAUSE TestAccDialogflowCXEnvironment_update
=== RUN   TestAccDialogflowCXFlow_update
=== PAUSE TestAccDialogflowCXFlow_update
=== RUN   TestAccDialogflowCXIntent_update
=== PAUSE TestAccDialogflowCXIntent_update
=== RUN   TestAccDialogflowCXPage_update
=== PAUSE TestAccDialogflowCXPage_update
=== RUN   TestAccDialogflowCXVersion_update
=== PAUSE TestAccDialogflowCXVersion_update
=== CONT  TestAccDialogflowCXAgent_dialogflowcxAgentFullExample
=== CONT  TestAccDialogflowCXEntityType_update
=== CONT  TestAccDialogflowCXPage_update
=== CONT  TestAccDialogflowCXIntent_dialogflowcxIntentFullExample
=== CONT  TestAccDialogflowCXVersion_update
=== CONT  TestAccDialogflowCXFlow_update
=== CONT  TestAccDialogflowCXVersion_dialogflowcxVersionFullExample
=== CONT  TestAccDialogflowCXIntent_update
2022/04/12 10:51:36 [DEBUG] [transport] transport: error closing conn during Close: close unix /var/folders/jf/b6pmcvvn6qb8bp5g7myw_s7800jyp0/T/plugin108296218->: use of closed network connection
--- PASS: TestAccDialogflowCXAgent_dialogflowcxAgentFullExample (14.03s)
=== CONT  TestAccDialogflowCXAgent_update
--- PASS: TestAccDialogflowCXIntent_dialogflowcxIntentFullExample (14.45s)
=== CONT  TestAccDialogflowCXEnvironment_dialogflowcxEnvironmentFullExample
--- PASS: TestAccDialogflowCXVersion_dialogflowcxVersionFullExample (70.66s)
=== CONT  TestAccDialogflowCXFlow_dialogflowcxFlowFullExample
--- PASS: TestAccDialogflowCXFlow_dialogflowcxFlowFullExample (9.63s)
=== CONT  TestAccDialogflowCXPage_dialogflowcxPageFullExample
--- PASS: TestAccDialogflowCXIntent_update (150.43s)
=== CONT  TestAccDialogflowCXEnvironment_update
=== CONT  TestAccDialogflowCXEntityType_dialogflowcxEntityTypeFullExample
--- PASS: TestAccDialogflowCXVersion_update (150.54s)
--- PASS: TestAccDialogflowCXEntityType_dialogflowcxEntityTypeFullExample (10.10s)
--- PASS: TestAccDialogflowCXPage_dialogflowcxPageFullExample (81.23s)
--- PASS: TestAccDialogflowCXPage_update (177.03s)
--- PASS: TestAccDialogflowCXFlow_update (177.18s)
--- PASS: TestAccDialogflowCXAgent_update (163.27s)
--- PASS: TestAccDialogflowCXEntityType_update (184.55s)
--- PASS: TestAccDialogflowCXEnvironment_dialogflowcxEnvironmentFullExample (189.25s)
--- PASS: TestAccDialogflowCXEnvironment_update (151.52s)
PASS
ok  	github.com/hashicorp/terraform-provider-google-beta/google-beta	303.695s
```

The error in the middle is unusual, but doesn't seem to have caused any issues. This is a sample test file: https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/resource_dialogflowcx_environment_test.go

We do use ExternalProviders in the provider tests, but not all of them. I don't believe we use any in this set of tests

**Comment by bflad** [comment]
Hi again @slevenick 👋  Thank you so much for the additional information, very helpful for troubleshooting this. Out of curiosity, could you let someone know on the HashiCorp side (not through GitHub 😄 ) which GPG key you used to encrypt those files? Luckily the Security team here was able to handle it for me, but I just wanted to be sure there wasn't a special key floating around that others should know about.

My hunch here is that this may be _unfortunately_ somewhat expected behavior given the current acceptance testing logging flow. Essentially what happens is that for every call to `helper/resource.Test()` (what also underlies `helper/resource.ParallelTest()` and `helper/resource.UnitTest()`, one of the first things that happens is that we inspect the logging environment variables (e.g. `TF_LOG_PATH_MASK`) to determine if we should override the default standard library `log` package logger. Since these tests are being executed in parallel, their startup is randomized and that override is happening based each new test starting up, which would set the default logger to the latest test name started up at the time. Any logs going through the default logger would wind up there until the next test starts and changes it again.

The `TF_LOG_PATH_MASK` logging setup worked _really_ well for us in the TeamCity environment because each test was actually spun up in its own separate `go test` process (if you're curious, via [`teamcity-go-test`](https://github.com/jen20/teamcity-go-test)) so we could separately capture panics and continue running other tests without issues. Overriding the default standard library `log` package logger in that case didn't matter, since it was only affecting each individual test. It does however make a regular, parallelized `go test` run have this particular issue.

Trying to fix this is an interesting challenge! Disabling the parallelism _should_ resolve the issue (it'd be super awesome if you could confirm that by running your `go test` with `-parallel=1` or switching things to `helper/resource.Test()`), but that is certainly less than ideal.

What is more likely necessary to happen is that whatever logging in your provider/testing either needs to:

- Create your own separate logger for each test that mimics the startup environment variable lookup of the testing framework and terraform-plugin-log, then have all of your logging code set to use that new logger for the rest of the test
- Funnel logs through terraform-plugin-log's logging system, which ensures there are separate loggers per test. This should work well to convert provider logging code if it is calling regular `log.*` calls today. If there is another logging library in use, we will need to enhance terraform-plugin-log so you can give the logging library our `io.Writer`. One place that will require some  design work would be making it easier for providing the correct logging pieces for _testing_ code outside `helper/resource.Test()`, as we'd likely need to create a new terraform-plugin-log function to give an `io.Writer` outside setting up its usual `context.Context` based setup.

Does this make sense? Happy to answer questions and we can figure out next steps together.

**Comment by bflad** [comment]
Oh and for what its worth, terraform-plugin-sdk/v2@2.14.0 should help with making any Terraform CLI and `ExternalProviders` logging now be in the correct place when using `TF_LOG_PATH_MASK`.

**Comment by slevenick** [comment]
Ok we will try a wrapper. Our current solution is to limit parallelization when we are attempting to capture accurate logs, but that's obviously pretty slow
